xcdat/temporal_avg.py

```diff
@@ -0,0 +1,950 @@
+            Center the time coordinate using time bounds, by default True.
+        season_config: SeasonConfig, optional
+            A dictionary for "season" frequency configurations. If configs for
+            predefined seasons are passed, configs for custom seasons are
+            ignored and vice versa.
+
+            Configs for predefined seasons:
+
+            * "dec_mode" (Literal["DJF", "JFD"], by default "DJF")
+               The mode for the season that includes December.
+
+                * "DJF": season includes the previous year December.
+                * "JFD": season includes the same year December. Xarray
+                  incorrectly labels the season with December as "DJF" when it
+                  should be "JFD". Refer to [1]_ for more information on this
+                  xarray behavior.
+
+            * "drop_incomplete_djf" (bool, by default True)
+                If the "dec_mode" is "DJF", this flag drops (True) or keeps
+                (False) time coordinates that fall under incomplete DJF seasons
+                Incomplete DJF seasons include the start year Jan/Feb and the
+                end year Dec.
+
+            Configs for custom seasons:
+
+            * "custom_seasons" ([List[List[str]]], by default None)
+                List of sublists containing month strings, with each sublist
+                representing a custom season.
+
+                * Month strings must be in the three letter format (e.g., 'Jan')
+                * Each month must be included once in a custom season
+                * Order of the months in each custom season does not matter
+                * Custom seasons can vary in length
+
+                >>> # Example of custom seasons in a three month format:
+                >>> custom_seasons = [
+                >>>     ["Jan", "Feb", "Mar"],  # "JanFebMar"
+                >>>     ["Apr", "May", "Jun"],  # "AprMayJun"
+                >>>     ["Jul", "Aug", "Sep"],  # "JunJulAug"
+                >>>     ["Oct", "Nov", "Dec"],  # "OctNovDec"
+                >>> ]
+
+        Returns
+        -------
+        xr.Dataset
+            Dataset containing the averaged data variable.
+
+        References
+        ----------
+        .. [1] https://github.com/pydata/xarray/issues/810
+
+        Examples
+        --------
+        Import temporal averaging functionality:
+
+        >>> import xcdat
+
+        Access temporal averaging method:
+
+        >>> # First option
+        >>> ds.temporal.temporal_avg(...)
+        >>> # Second option
+        >>> ds.xcdat.temporal_avg(...)
+
+        Get weighted temporal average for a variable:
+
+        >>> ds_year = ds.temporal.temporal_avg(
+        >>>     "timeseries",
+        >>>     "year",
+        >>>     data_var="ts"
+        >>> )
+        >>> ds_year.ts
+
+        Get weighted temporal average for a variable ("season" frequency):
+
+        >>> ds_season_with_djf = ds_season.temporal.temporal_avg(
```

Comment:
I'm confused by the dataset being passed in. Should this be: `ds_season_with_djf = ds.temporal.temporal_avg(`? Or should I have done some work on the original raw dataset `ds`?

Reply:
Another issue with seasonal frequency (but not month/annual) is that I am getting the following error:
```
fn = '/p/css03/esgf_publish/CMIP6/CMIP/NCAR/CESM2/historical/r1i1p1f1/Amon/tas/gn/v20190308/tas_Amon_CESM2_historical_r1i1p1f1_gn_185001-201412.nc'
ds = xcdat.open_dataset(fn)
ds = ds.xcdat.temporal_avg("timeseries", "season")
```

> TypeError: '_ElementwiseFunctionArray' object does not support item assignment

Reply:
> I'm confused by the dataset being passed in. Should this be: `ds_season_with_djf = ds.temporal.temporal_avg(`? Or should I have done some work on the original raw dataset `ds`?

That is a typo, it should be `ds_season_with_djf = ds.temporal.temporal_avg(...`.


> Another issue with seasonal frequency (but not month/annual) is that I am getting the following error:
> 
> ```
> fn = '/p/css03/esgf_publish/CMIP6/CMIP/NCAR/CESM2/historical/r1i1p1f1/Amon/tas/gn/v20190308/tas_Amon_CESM2_historical_r1i1p1f1_gn_185001-201412.nc'
> ds = xcdat.open_dataset(fn)
> ds = ds.xcdat.temporal_avg("timeseries", "season")
> ```
> 
> > TypeError: '_ElementwiseFunctionArray' object does not support item assignment

I opened up an issue in the xarray repo (https://github.com/pydata/xarray/issues/6015) and thought I implemented a workaround for this using `.load()` on the time bounds. I'll take a look, thanks for the example Steve.

Reply:
You may be on an outdated version of the branch. I ran the same example on the latest commit and it works.